### PR TITLE
Continue sidebar menu grouping feature

### PR DESCRIPTION
**Pull request by adam-code-labx**
@glennjacobs I can see this one was started so I have continued this grouping feature.
If we can't make the next beta release then please include in the new develop branch to avoid switching branches 😂

### CODE REVIEW

File: packages/admin/resources/views/components/menu-list.blade.php [hunk 1]
```diff
@@ -25,22 +37,36 @@
...
         @if (count($section->getItems()))
             <div>
-                <header class="text-sm font-semibold text-gray-600">
+                <header class="text-sm font-semibold text-gray-600"
+                        x-cloak
+                        x-show="type == 'main_menu' && showExpandedMenu || type == 'sub_menu'">
                     {{ $section->name }}
                 </header>
 
-                <ul class="mt-2 space-y-2">
```
Comment (markmead) [inline packages/admin/resources/views/components/menu-list.blade.php:32]: Removing `mt-2` has caused the menu to be squashed with the menu title.

File: packages/admin/resources/views/components/menu-list.blade.php [hunk 2]
```diff
@@ -1,18 +1,30 @@
-<div class="space-y-4">
+<div class="space-y-4" x-data="{ type: '{{ $type }}' }">
```
Comment (markmead) [inline packages/admin/resources/views/components/menu-list.blade.php:1]: I think we should be more declarative here, `type` doesn't mean much. Switching to `menuType` would help.

File: packages/admin/resources/views/components/menu-list.blade.php [hunk 3]
```diff
@@ -25,22 +37,36 @@
...
+                        x-cloak
+                        x-show="type == 'main_menu' && showExpandedMenu || type == 'sub_menu'">
                     {{ $section->name }}
                 </header>
 
-                <ul class="mt-2 space-y-2">
+                <ul class="ml-1 space-y-2 flex flex-col">
                     @foreach ($section->getItems() as $item)
-                        <li>
+                        <li :class="{ 'items-center': !showExpandedMenu }">
```
Comment (markmead) [inline packages/admin/resources/views/components/menu-list.blade.php:48]: Do you have a screenshot of how this looks? 📸

Comment (adam-code-labx) [reply packages/admin/resources/views/components/menu-list.blade.php:48]: sure 👍 
Screenshot 2022-07-27 at 10 37 58

Comment (markmead) [reply packages/admin/resources/views/components/menu-list.blade.php:48]: I could be dumb here but that doesn't look much different from the current implementation without `:class="{ 'items-center': !showExpandedMenu }"`

File: packages/admin/resources/views/livewire/sidebar.blade.php [hunk 4]
```diff
@@ -1,41 +1,11 @@
...
-                              x-show="showTooltip"
-                              class="absolute z-10 p-2 ml-4 text-xs text-center text-white bg-gray-900 rounded dark:bg-gray-800 w-28 left-full">
-                            {{ $item->name }}
-                        </span>
-                    </a>
-                </li>
-            @endforeach
-        </ul>
+        <x-hub::menu-list
+                menuType="main_menu"
```
Comment (markmead) [inline packages/admin/resources/views/livewire/sidebar.blade.php:5]: I'm being picky here (sorry) but could this be passed through as `menu-type`, I believe you won't need to change the `$menuType` used elsewhere.

File: packages/admin/resources/views/components/menu-list.blade.php (hunk 3 above)
Comment (adam-code-labx) [reply packages/admin/resources/views/components/menu-list.blade.php:48]: Your right no difference at all will remove it, in my defence I copied this over from `sidebar.blade.php` 😬

File: packages/admin/resources/views/components/menu-list.blade.php [hunk 5]
```diff
@@ -1,18 +1,30 @@
...
     @if (count($items))
         <div>
-            <ul class="space-y-2">
+            <ul class="ml-1 space-y-2">
                 @foreach ($items as $item)
                     <li>
                         <a href="{{ route($item->route) }}"
+                           x-data="{ showTooltip: false }"
+                           x-on:mouseover="showTooltip = showExpandedMenu ? false : true"
+                           x-on:mouseleave="showTooltip = false"
```
Comment (markmead) [inline packages/admin/resources/views/components/menu-list.blade.php:10]: This approach was switched out to use the `group` class in Tailwind CSS -https://github.com/getcandy/getcandy/blob/develop/packages/admin/resources/views/livewire/sidebar.blade.php

File: packages/admin/resources/views/components/menu-list.blade.php [hunk 6]
```diff
@@ -1,18 +1,30 @@
-<div class="space-y-4">
+<div class="space-y-4" x-data="{ menuType: '{{ $menuType }}' }">
     @if (count($items))
         <div>
-            <ul class="space-y-2">
+            <ul class="ml-1 space-y-2">
```
Comment (markmead) [inline packages/admin/resources/views/components/menu-list.blade.php:4]: Why the addition of `ml-1`?

**Review by markmead — CHANGES_REQUESTED** [review]
LGTM! Just a few small tweaks where the branch has got outdated.

File: packages/admin/resources/views/components/menu-list.blade.php (hunk 5 above)
Comment (adam-code-labx) [reply packages/admin/resources/views/components/menu-list.blade.php:10]: No problem will check this one over the weekend and update.

File: packages/admin/resources/views/components/menu-list.blade.php (hunk 6 above)
Comment (adam-code-labx) [reply packages/admin/resources/views/components/menu-list.blade.php:4]: Good question it was almost a month ago so will check once I have merged  the latest changes 😆

Comment (adam-code-labx) [reply packages/admin/resources/views/components/menu-list.blade.php:4]: The reason. ml-1 was added was to make sure the menu items line up with the dashboard.

Screenshot 2022-08-27 at 10 54 39

**Comment by adam-code-labx** [comment]
@markmead Tooltip updated to group hover.

I have quite a few ideas on improving this menu grouping feature.
Already implemented part of this for my client customisations the ability to dynamically populate the menu from the page components.

With this I am introducing a new Resource class to handle the menu and PageComponent to handle the view.
The page CRUD components extend the resource so only need to define the navigation once.

```php
<?php

namespace GetCandy\Hub\Http\Livewire\Pages\Products;

use GetCandy\Hub\Http\Livewire\PageComponent;

abstract class ProductsResource extends PageComponent
{
    protected static string $navigationLabel = 'Products';

    protected static string $navigationIcon = 'shopping-cart';

    protected static string $navigationGroup = 'catalogue';
    
    protected static int $navigationPosition = 3;

    public function mount()
    {
        static::setTitle('Products');
    }
}
```

```php
<?php

namespace GetCandy\Hub\Http\Livewire;

use Livewire\Component;

class PageComponent extends Component
{
    protected static string $view;

    protected static string $navigationLabel;

    protected static string $navigationIcon;

    protected static string $navigationGroup;
    
    protected static int $navigationPosition;

    protected static array $params = [];

    public static function setTitle(string $title)
    {
        self::$params['title'] = $title ?? 'Page';
    }

    /**
     * @param  array  $params
     */
    public static function setParams(array $params): void
    {
        self::$params = array_merge(self::$params, $params);
    }

    /**
     * Render the livewire component.
     *
     * @return \Illuminate\View\View
     */
    public function render()
    {
        return view('adminhub::livewire.pages.'.static::$view)
            ->layout('adminhub::layouts.app', static::$params);
    }
}
```

This is how it looks currently on my side.
Screenshot 2022-08-27 at 11 04 03

If your interested in including this let me know, I can either refactor the existing page components or create another PR for this one 😜

**Comment by glennjacobs** [comment]
@adam-code-labx got a few bugs to sort on this one.

Firstly, the styling seems to have gone a bit wrong in places.

image

The tooltips do not align correctly and are inconsistent. And the icon placement is not centered in the light blue boxes.

Also, it no longer remembers your preference when contracting the menu. If you refresh it will forget and go back to default (expanded).

**Comment by adam-code-labx** [comment]
@markmead I was about to push a fix but now lots of merge conflicts 😂

**Comment by adam-code-labx** [comment]
@markmead you can review here then apply the changes just put up temp branch, this is the only commit for the changes.
https://github.com/adam-code-labx/getcandy/commit/63901694cc392cd53793cbc293c1d13a30701763

Currently you do have an issue with the setting alignment should only be applied to the main menu items
Screenshot 2022-09-05 at 12 31 50

Also the focus when in the settings menu should be only the settings menu see my collapsedMainMenu override, which will ignore persist if you are in a settings section.

**Comment by markmead** [comment]
Sorry, @adam-code-labx I was asked to take a quick look as we're getting PRs merged in today!

Welcome to make those changes/fixes, but we want a working persist more than we want the settings menu being displayed by default when on `/settings`.

**Comment by markmead** [comment]
Looks good! @adam-code-labx

Do need to run `yarn production` as the JS is now 37k lines long 😅 

Question for @glennjacobs do we want the side menu auto collapsing when a user is on `/hub/settings/*`?
Screenshot 2022-09-06 at 07 42 55

If we do take that approach, should the same happen for collections and variants?
Screenshot 2022-09-06 at 07 44 46
Screenshot 2022-09-06 at 07 43 49

Additionally, "Orders List" does that sound right? I feel like it should be "Orders" but then you have the issue of the sub-menu header being the same. Looking at what Shopify does, "Orders" is a link and the sub-menu header.
Screenshot 2022-09-06 at 07 47 59

**Comment by adam-code-labx** [comment]
Ooops that’s what happens with `npm run watch` 😆

**Comment by adam-code-labx** [comment]
@markmead @glennjacobs collapsing the main menu if there is a sub menu open is a great idea to focus on the section you are on.

For orders I would change the group name to be something like Sales which would eventually have the following:

Sales
  - Customers
  - Discounts
  - Orders 
  - Carts

**Comment by adam-code-labx** [comment]
I feel like soon we will need to consider accordion style collapsible especially once we have some addons in place.

**Comment by glennjacobs** [comment]
> Question for @glennjacobs do we want the side menu auto collapsing when a user is on /hub/settings/*?

I think we can, but only for smaller screen sizes. If there is enough room, it should remain expanded (if that was their preference).

**Comment by adam-code-labx** [comment]
Ok so I should revert the settings commit?
Then just let the user decide? 

Let me know if you would like any further changes or @markmead you plan to finish this one?

**Comment by glennjacobs** [comment]
> For orders I would change the group name to be something like Sales which would eventually have the following:
> 
> Sales
> 
> * Customers
> * Discounts
> * Orders
> * Carts

I think "Sales" works, go with that.

**Comment by markmead** [comment]
All good for you to finish this one up 🏁 @adam-code-labx

**Comment by adam-code-labx** [comment]
@markmead @glennjacobs 
Ok should be good to merge now.

**Comment by jonathangustafssonax** [comment]
@markmead is there any blocker to merge this in to main?

**Comment by glennjacobs** [comment]
> @markmead is there any blocker to merge this in to main?

We have been testing different menu layouts internally. This is our preference currently. 

https://ui-menu.vercel.app/

Ignore the style, it's more about how the menu works. So you'll see the sections are actually replaced with top-level menu items. We are likely to take this route, but was keeping this PR open until we made a firm decision.

**Comment by glennjacobs** [comment]
Closing in favour of #680

**Pull request closed without merging**